colossus/src/main/scala/colossus/protocols/http/StreamService.scala

```diff
@@ -0,0 +1,222 @@
+}
+
+case class StreamingHttpRequest(head: HttpRequestHead, body: Source[Data]) extends StreamingHttpMessage[HttpRequestHead]
+case class StreamingHttpResponse(head: HttpResponseHead, body: Source[Data]) extends StreamingHttpMessage[HttpResponseHead]
+
+
+trait StreamingHttp extends Protocol {
+  type Request = StreamingHttpMessage[HttpRequestHead]
+  type Response = StreamingHttpMessage[HttpResponseHead]
+}
+
+
+object GenEncoding {
+
+  type StreamHeader = Protocol {
+    type Request = HttpRequestHead
+    type Response = HttpResponseHead
+  }
+
+  type HeadEncoding = Encoding {
+    type Input <: HttpMessageHead
+    type  Output <: HttpMessageHead
+  }
+
+  type GenEncoding[M[T <: HttpMessageHead], E <: HeadEncoding] = Encoding {
+    type Input = M[E#Input]
+    type Output = M[E#Output]
+  }
+
+}
+import GenEncoding._
+
+
+trait InputMessageBuilder[T <: HttpMessageHead] {
+  def build(head: T): (Sink[Data], StreamingHttpMessage[T])
+}
+object StreamRequestBuilder extends InputMessageBuilder[HttpRequestHead] {
+  def build(head: HttpRequestHead) = {
+    val pipe = new BufferedPipe[Data](10)
+    (pipe, StreamingHttpRequest(head, pipe))
+  }
+}
+
+class StreamServiceServerController[E <: HeadEncoding](
+  val downstream: ControllerDownstream[GenEncoding[StreamingHttpMessage, E]],
+  builder: InputMessageBuilder[E#Input]
+)
+extends ControllerDownstream[GenEncoding[HttpStream, E]] 
+with DownstreamEventHandler[ControllerDownstream[GenEncoding[StreamingHttpMessage, E]]] 
+with ControllerUpstream[GenEncoding[StreamingHttpMessage, E]]
+with UpstreamEventHandler[ControllerUpstream[GenEncoding[HttpStream, E]]] {
+
+  downstream.setUpstream(this)
+
+  type InputHead = E#Input
+  type OutputHead = E#Output
+
+  private var currentInputStream: Option[Sink[Data]] = None
+
+  private val outputStreams = new MessageQueue[Source[HttpStream[OutputHead]]](100)
+
+  protected def fatal(message: String) {
+    println(s"FATAL ERROR: $message")
+    upstream.connection.forceDisconnect()
+  }
+
+  def processMessage(input: HttpStream[InputHead]) {
+    input match {
+      case Head(head) => currentInputStream match {
+        case None => {
+          val (sink, msg) = builder.build(head)
+          currentInputStream = Some(sink)
+          downstream.processMessage(msg)
+        }
+        case Some(uhoh) => {
+          //we got a head before the last stream finished, not good
+          fatal("received head during unfinished stream")
+        }
+      }
+      case b @ Data(_, _) => currentInputStream match {
+        case Some(sink) => sink.push(b) match {
+          case PushResult.Filled(signal) => {
+            upstream.pauseReads()
+            signal.react{
+              upstream.resumeReads()
+            }
+          }
+          case bad: PushResult.NotPushed => {
+            // :(
+            fatal("failed to push message to stream")
+          }
+          case other => {}
```

Comment:
actually in this case there is no error and no further action needs to be taken.  I'll add some comments to make that more clear